README
About
=====================================
open flash chart 2 plugin for rails applications

This library was ported from the open flash chart project's php code to be used with Ruby on Rails.
Visit the open flash chart project here - http://teethgrinder.co.uk/open-flash-chart-2/

NOTE:
sample application is using haml
  gem install haml
plugin is using json
  gem install json

Details
=====================================
Go to http://openflashchart2.sentor.pl for details

Install plugin
=====================================
get only plugin to existing application:
go to plugins directory
    cd vendor/plugins
clone git repo to open_flash_chart_2 directory
    git clone git://github.com/korin/open_flash_chart_2_plugin.git open_flash_chart_2
go to main rails directory
    cd ../..
install it using rake command
    rake open_flash_chart_2:install

Example of usage
=====================================
include js file in loyout or anything else where you need that:
  = javascript_include_tag 'swfobject.js'

in some conroller:
  class SomeController < ApplicationController
    def funny_action
                   #width, height, link to data
      @graph = ofc2(650,300, "charts_ofc2/line") 
    end
  end

in charts_ofc2 conroller:
  class ChartsOfc2Controller < ApplicationController
    def line
      title = OFC2::Title.new( DateTime.now.strftime('%Y-%m-%d %H:%M'), "{font-size: 14px; color: #b50F0F; text-align: center;}")
      line_dot = OFC2::Line.new
      line_dot.values= [9,8,7,6,5,4,3,2,1]
      chart = OFC2::Graph.new
      chart.title= title
      chart << line_dot
      render :text => chart.render
    end
  end

in views/some/funny_action.haml view
  = @graph

That's it, try it.
If You have trouble try sample application (info below), or ask me va e-mail (lisukorin@gmail.com).


Get sample application
=====================================
Grab whole sample application:
go to directory where you want create sample application
    git clone git://github.com/korin/open_flash_chart_2.git
install plugin (instruction is above)
customize database.yml
customize what you want ;)

Help
=====================================
Looking for help?
try:
home page: http://openflashchart2.sentor.pl/infos
or email to me: lisukorin@gmail.com,
  don't forget write 'open flash chart' in subject or my mail client wil treat your message as spam.



CHANGES
[0.9] rc-1 : 15.04.2009
  - remove to_json method from OWJSON module
  - remove remove require 'json' from plugin
  - migrate sample application to rails 2.3
  - remove gem 'json' from enviroment.rb
[0.8] rc-1 : 09.01.2009
  - add shape drawning support, example is here:
     http://openflashchart2.sentor.pl/chart_examples/mix_advanced_draw_shape
  - add stacked bar keys support:
     http://openflashchart2.sentor.pl/chart_examples/stacked_bar
  - pseudo horizontal stacked bar, example:
     http://openflashchart2.sentor.pl/chart_examples/mix_advanced_shapes
[0.7] rc-1 : 12.12.2008
  - add Tooltip class, example is here:
     http://openflashchart2.sentor.pl/chart_examples/mix_advanced_tooltip
[0.6] rc-1 : 11.12.2008
  - now generating many inline chart on one page is possible
  - moved to open-flash-chart-2-hyperion version
  - remove spelling errors
  - remove set_tooltip method, use instead set_tip or tip=
[0.5] beta-4
  - methods are now generated by method_missing
[0.4] beta-3
  - add ofc2_inline method to generate graph without url
  - add instance variables getters, changes proposed by __Pavlo Masko__
[0.3] beta-2
  - changed installation instructions
  - add rake tasks to install/uninstall
[0.2] beta-1
  - change README - add description, install section and info about contact
  - make examples with basic charts
  - compare plugin with php library and add missing class/methods
[0.1] alpha
  - rewrite php library


lib/open_flash_chart_2.rb
module OFC2
  # specjal module included in each class
  # with that module we add to_hash and to_json methods
  # there is also a method_missing which allow user to set/get any instance variable
  # if user try to get not setted instance variable it return nil and generate a warn
  module OWJSON

    # return a hash of instance values
    def to_hash
      self.instance_values
    end
    alias :to_h :to_hash

    # if You use rails older that 2.3  probably  you have to uncomment that method and add require 'json' in enviroment.rb
    # otherwise to_json method will not work propertly
    #    # You can pass options to to_json method, but remember that they have no effects!!!
    #    # argument 'options' is for rails compability
    #    def to_json(options = {})
    #      to_hash.to_json
    #    end

    # method_missing handle setting and getting instance variables
    # You can set variable in two ways:
    # 1. variable_name = value
    # 1. set_variable_name(value)
    # you can get only  alredy setted variables, otherwise warnig is logged and return nil
    def method_missing(method_id, *arguments)
      a = arguments[0] if arguments and arguments.size > 0
      method = method_id.to_s
      if method =~ /^(.*)(=)$/
        self.instance_variable_set("@#{$1.gsub('_','__')}", a)
      elsif method =~ /^(set_)(.*)$/
        self.instance_variable_set("@#{$2.gsub('_','__')}", a)
      elsif self.instance_variable_defined?(method_id)
        self.instance_variable_get("@#{method_id.to_s.gsub('_','__')}") # that will be return instance variable value or nil, handy
      else
        #        well there is no instance variable and user don't wan't to define any
        #         maybe better return nil and warn that rise exception(call super) ?
        warning = <<-EOF
          !!! there is no instance variable named #{method_id} !!!
          - if You want to set instance variable use variable= or set_variable(var) methods
          - if You want to get variable call object for variable: obj.variable
          - You can call only for variables You set before
        EOF
        warn(warning)
        nil
      end
    end
  end

  # include methods to controller
  def self.included(controller)
    controller.helper_method(:ofc2, :ofc2_inline)
  end

  # generate a ofc object using Graph object
  #  +width+ width for div
  #  +height+ height for div
  #  +graph+ a OFC2::Graph object
  #  +base+ uri for graph, default '/', not used in this method, go to ofc2 method for details
  #  +id+ id for div with graph, default Time.now.usec
  #  +swf_base+ uri for swf file, default '/'
  def ofc2_inline(width, height, graph, base='/', id=Time.now.usec, swf_base='/')
    div_name = "flashcontent_#{id}"
    <<-EOF
      <div id="#{div_name}"></div>
      <script type="text/javascript">

        function #{div_name}_data(){
          return '#{graph.render}';
        };

        // i'm not shure that is necessary
        function findSWF(movieName) {
          if (navigator.appName.indexOf("Microsoft")!= -1) {
            return window[movieName];
          } else {
            return document[movieName];
          }
        };

        swfobject.embedSWF(
          '#{swf_base}open-flash-chart.swf', '#{div_name}',
          '#{width}', '#{height}','9.0.0', 'expressInstall.swf',
          {'get-data':'#{div_name}_data'} );

      </script>
    EOF
  end

  # generate a ofc object using data from url
  #  +width+ width for div
  #  +height+ height for div
  #  +url+ an url which return data in json format, if you use url_for method to set url param the base param must be set to '' (empty string)
  #  +base+ uri for graph, default '/'
  #  +id+ id for div with graph, default Time.now.usec
  #  +swf_base+ uri for swf file, default '/'
  def ofc2(width, height, url, base='/', id =Time.now.usec, swf_base='/')
    url = CGI::escape(url)
    div_name = "flashcontent_#{id}"
    <<-EOF
      <div id='#{div_name}'></div>
      <script type="text/javascript">
        swfobject.embedSWF(
        "#{swf_base}open-flash-chart.swf","#{div_name}",
        "#{width}", "#{height}", "9.0.0", "expressInstall.swf",
        {"data-file":"#{base}#{url}"} );
      </script>
    EOF
  end

  # insance variables:
  #
  #  +style+ style for element, it's is in css style eg. "{font-size: 20px; color: #FF0F0F; text-align: center;}"
  #  +text+ text for element
  class Element
    include OWJSON
    # You can initialize Elemnt while creating it, otherwise it will be have default valiues
    #  +text+ = ''
    #  +css+ = "{font-size: 20px; color: #FF0F0F; text-align: center;}"
    def initialize(text = '', css = "{font-size: 20px; color: #FF0F0F; text-align: center;}")
      @text= text
      @style= css
    end
  end

  # documentation is the same as Element class
  class XLegend <  Element ;end
  # documentation is the same as Element class
  class Title < Element ;end
  # documentation is the same as Element class
  class YLegend <  Element ;end

  # YAxisBase
  #
  #  +stroke+
  #  +tick_length+
  #  +colour+
  #  +min+
  #  +max+
  #  +steps+
  #  +labels+
  class YAxisBase
    include OWJSON

    # set colour and grid_colour at once
    # there is also an alias colours=
    #
    #  +colour+ colour for labels eg. #FF0000
    #  +grid_colour+ colour for grid eg. #00FF00
    def set_colours( colour, grid_colour )
      set_colour( colour )
      set_grid_colour( grid_colour )
    end
    alias_method :colours=, :set_colours


    # set range at once
    # there is also an alias range=
    #
    #  +min+ minimum for y_axis
    #  +max+ maximum for y_axis
    #  +steps+ how many steps skip before print label
    def set_range( min, max, steps=1 )
      set_min(min)
      set_max(max)
      set_steps( steps )
    end
    alias_method :range=, :set_range

    # set offset for axis, these is handy when You want to 3d graph render
    # there is also an alias offset=
    def set_offset( off )
      @offset = off ? 1 : 0
    end
    alias_method :offset=, :set_offset

  end

  class YAxis < YAxisBase
    #    # left axis control grid colour, but right not
    #    def set_grid_colour(color = '#ff0000')
    #      @grid__colour = color
    #    end
    #    alias_method :grid_colour=, :set_grid_colour
  end
  class YAxisRight < YAxisBase
  end

  # x_axis
  #
  #  +stroke+
  #  +tick_length+
  #  +colour+
  #  +tick_height+
  #  +grid_colour+
  #  +min+
  #  +max+
  #  +steps+
  #  +labels+
  #  +offset+
  class XAxis
    include OWJSON

    # well it must be done this way because instance variable name can't be started by number
    %w(3d).each do |method|
      define_method("set_#{method}") do |a|
        self.instance_variable_set("@___#{method}", a)
      end
      define_method("_#{method}=") do |a|
        self.instance_variable_set("@___#{method}", a)
      end
      define_method("#{method}") do
        self.instance_variable_get("@___#{method}")
      end
    end
    # set +colour+ and +grid_colour+, use a css color style '#ff00ff'
    def set_colours( colour, grid_colour )
      @colour= colour
      @grid_colour= grid_colour
    end

    # o is treat as a logic
    def set_offset( o )
      @offset = !!o
    end

    # helper method to make the examples simpler.
    def set_labels_from_array( a )
      x_axis_labels = XAxisLabels.new
      x_axis_labels.set_labels( a )
      x_axis_labels.set_steps( @steps ) if @steps
      @labels = x_axis_labels
    end
    alias_method :labels_from_array=, :set_labels_from_array
    def set_range( min, max )
      @min=min
      @max=max
    end
  end

  #  +text+
  #  +colour+
  #  +size+
  #  +rotate+
  #  +visible+
  class XAxisLabel
    include OWJSON
    def initialize( text = nil, colour = nil, size= nil)
      @text= text if text
      @colour= colour if colour
      @size= size if size
    end
    def set_vertical
      @rotate = "vertical"
    end
    alias_method :vertical, :set_vertical

    def set_horizontal
      @rotate = "horizontal"
    end
    alias_method :horizontal, :set_horizontal
  end

  #  +steps+
  #  +labels+
  #  +colour+
  #  +size+
  class XAxisLabels
    include OWJSON
    def set_vertical()
      @rotate = "vertical"
    end
  end

  # scatter value
  #
  #  +x+
  #  +dot_size+
  #  +y+
  class ScatterValue
    include OWJSON
    def initialize( x, y, dot_size=-1 )
      @x = x
      @y = y
      set_dot_size(dot_size) if dot_size > 0
    end
  end

  #  +colour+
  #  +dot_size+
  #  +values+
  class Scatter
    include OWJSON
    def initialize( colour, dot_size )
      @type      = "scatter"
      set_colour( colour )
      set_dot_size( dot_size )
    end
  end



  #  +title+
  #  +x_axis+
  #  +y_axis+
  #  +y_axis_right+
  #  +x_legend+
  #  +y_legend+
  #  +bg_colour+
  #  +elements+
  class Graph
    include OWJSON

    # it must be done in that way because method_missing method replace _ to __,
    # maybe I add seccond parameter to handle with that
    %w(x_axis y_axis y_axis_right x_legend y_legend bg_colour).each do |method|
      define_method("set_#{method}") do |a|
        self.instance_variable_set("@#{method}", a)
      end
      define_method("#{method}=") do |a|
        self.instance_variable_set("@#{method}", a)
      end
      define_method("#{method}") do
        self.instance_variable_get("@#{method}")
      end
    end

    def initialize
      @title = Title.new( "Graph" )
      @elements = []
    end

    def add_element( e )
      @elements << e
    end
    alias_method :<<, :add_element

    def render
      s = to_json
      # everything about underscores
      s.gsub!('___','') # that is for @___3d variable
      s.gsub!('__','-') # that is for @smt__smt variables
      # variables @smt_smt should go without changes
      s
    end
  end

  # line chart
  #
  #  +values+
  #  +width+
  #  +colour+
  #  +font_size+
  #  +dot_size+
  #  +halo_size+
  #  +text+
  class LineBase
    include OWJSON
    def initialize(text = 'label text', font_size=10, values = [9,6,7,9,5,7,6,9,7], colour = '#0000FFs')
      @type      = "line_dot"
      @text      = text
      @font__size = font_size
      @values    = values
      @colour    = colour
    end
  end


  #  +value+
  #  +colour+
  #  +tip+
  #  +size+
  class DotValue
    include OWJSON
    def initialize(value = 0, colour = '', tip = nil)
      @value = value
      @colour = colour
      @tip = tip if tip
    end
  end
  # go to class LineBase for details
  class LineDot < LineBase ;end

  # go to class LineBase for details
  class Line < LineBase
    def initialize(text = 'label text', font_size=10, values = [9,6,7,9,5,7,6,9,7], colour = '#00FF00')
      super
      @type      = "line"
    end
  end

  # go to class LineBase for details
  class LineHollow < LineBase
    def initialize(text = 'label text', font_size=10, values = [9,6,7,9,5,7,6,9,7], colour = '#FF0000')
      super
      @type      = "line_hollow"
    end
  end

  #  +width+
  #  +color+
  #  +values+
  #  +dot_size+
  #  +text+
  #  +font_size+
  #  +fill_alpha+
  class AreaHollow
    include OWJSON
    def initialize(fill_alpha = 0.35, values = [])
      @type      = "area_hollow"
      set_fill_alpha  fill_alpha
      @values    = values
    end
  end

  #  +alpha+
  #  +colour+
  #  +values+
  #  +text+
  #  +font_size+
  class BarBase
    include OWJSON
    def initialize(values = [], text = '', size = 10)
      @values = values
      @text = text
      @font__size = size
    end
    def set_key( text, size )
      @text = text
      @font__size = size
    end
    def append_value( v )
      @values << v
    end
    alias_method :<<, :append_value
  end

  # go to class BarBase for details
  class Bar < BarBase
    def initialize(values = [], text = '', size = 10)
      super
      @type      = "bar"
    end
  end

  #  +top+
  #  +colour+
  #  +tip+
  class Value
    include OWJSON
    def initialize(top = 0, color = '', tip = nil)
      @top = top
      @colour = color
      @tip = tip
    end
  end
  class Bar3d < BarBase
    def initialize(values = [], text = '', size = 10)
      super
      @type      = "bar_3d"
    end
  end

  # go to class BarBase documentation for details
  class BarGlass < BarBase
    def initialize(values = [], text = '', size = 10)
      super
      @type      = "bar_glass"
    end
  end

  # go to class BarBase documentation for details
  #
  #  +offset+
  #  +colour+
  #  +outline_colour+
  class BarSketch < BarBase
    def initialize( colour = '#ff0000', outline_colour = '#00FF00', fun_factor = 5)
      @type      = "bar_sketch"
      set_colour( colour )
      set_outline_colour( outline_colour )
      @offset = fun_factor
    end
  end

  # go to class BarBase documentation for details
  # +keys+
  class BarStack < BarBase
    include OWJSON
    def initialize(values = [], text = '', size = 10, keys = [])
      super(values, text, size)
      @type      = "bar_stack"
      @keys      = keys
    end

    def append_key( key )
      @keys << key
    end

    def set_keys( keys )
      @keys = keys
    end
    alias_method :keys=, :set_keys


    def append_value( v )
      @values << v
    end
    alias_method :append_stack, :append_value
    alias_method :<<, :append_value
  end

  # go to class Value documentation for details
  #
  #  +val+
  #  +colour+
  class BarStackValue < Value
    include OWJSON
    def initialize(val, colour)
      @val = val
      @colour = colour
    end
  end

  #  +colour+
  #  +text+
  #  +font_size+
  class BarStackKey
    include OWJSON
    def initialize( colour, text, font_size )
      @colour = colour
      @text = text
      @font__size = font_size
    end
  end


  #  +left+
  #  +right+
  class HBarValue
    include OWJSON
    def initialize( left = 0, right = 0, tip = nil )
      @left = left
      @right = right
      @tip = tip if tip
    end
  end

  #  +colour+
  #  +text+
  #  +font_size+
  #  +values+
  class HBar
    include OWJSON
    def initialize(colour = "#9933CC", text = '', font_size = 10)
      @type      = "hbar"
      @colour    = colour
      @text      = text
      set_font_size font_size
      @values    = []
    end

    # v suppostu be HBarValue class
    def append_value( v )
      @values << v
    end
    alias_method :<<, :append_value

  end

  # pie value
  #  +value+
  #  +text+
  class PieValue
    include OWJSON
    def initialize( value, text )
      @value  = value
      @text   = text
    end
  end


  #  +colours+
  #  +alpha+
  #  +border+
  #  +values+
  #  +animate+
  #  +start_angle+
  class Pie
    include OWJSON
    def initialize(colours = ["#d01f3c","#356aa0","#C79810"],
        alpha = 0.6,
        border = 2,
        values = [2,3, PieValue.new(6.5, "hello (6.5)")]
      )
      @type     = 'pie'
      @colours  = colours
      @alpha	= alpha
      @border	= border
      @values	= values
    end
  end


  # +shadow+
  # +stroke+
  # +colour+ text colour
  # +background+ background colour
  # +title+ title style
  # +body+ body style
  class Tooltip
    include OWJSON

    def set_title_style( style = '')
      @title = style
    end

    def set_body_style( style = '')
      @body = style
    end

    def set_proximity
      @mouse = 1
    end
    alias_method :proximity, :set_proximity

    def set_hover
      @mouse = 2
    end
    alias_method :hover, :set_hover
  end


  class ShapePoint
    include OWJSON

    def initialize( x, y )
      @x = x
      @y = y
    end
  end

  class Shape
    include OWJSON
    def initialize( colour = '', values = [] )
      @type		= "shape"
      @colour	= colour
      @values	= values
    end

    def append_value( p )
      @values << p
    end
    alias_method :<<, :append_value
    alias_method :add_element, :append_value

  end

end
